flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/channel/ChannelStateWriterImpl.java

```diff
@@ -91,6 +91,7 @@ public ChannelStateWriterImpl(CheckpointStorageWorkerView streamFactoryResolver)
 
 	@Override
 	public void start(long checkpointId, CheckpointOptions checkpointOptions) {
+		results.keySet().forEach(oldCheckpointId -> abort(oldCheckpointId, new Exception("Starting new checkpoint " + checkpointId)));
```

Comment:
I guess the below `Preconditions.checkState(results.size() < maxCheckpoints, "results.size() > maxCheckpoints", results.size(), maxCheckpoints);` might be invalid if we remove the pending ones here?

Reply:
You are right that it's probably not needed. However, it's an inexpensive check that verifies the highlighted line.
I'd keep the check for now.